File: frontend/src/api/lifecycle.ts
import { API_BASE, apiFetch } from './http'

export interface StatusCatalogEntry {
  key: string
  label: string
  cdar_code: string
  mdt88_code: string | null
  manual_side: 'purchase' | 'sale' | null
  requires_detail: boolean
  requires_confirmation: boolean
}

export interface LifecycleCatalog {
  statuses: StatusCatalogEntry[]
  reasons: Record<string, string>
  actions: Record<string, string>
}

export interface LifecycleEventDetail {
  reason: string | null
  action: string | null
  comment: string | null
}

export interface LifecycleEventPayment {
  id: number
  amount: number
  currency: string
  payment_date: string
}

export interface LifecycleEventAttachment {
  id: number
  filename: string
  has_file: boolean
}

export interface LifecycleEventAfnorFlow {
  id: number
  flow_id: string | null
  direction: string
  flow_type: string
  syntax: string
  processing_rule: string | null
  state: string
  has_file: boolean
}

export interface LifecycleEvent {
  id: number
  invoice_id: number
  event_datetime: string
  status: string
  direction: string
  amount: number | null
  currency: string | null
  details: LifecycleEventDetail[]
  payments: LifecycleEventPayment[]
  attachments: LifecycleEventAttachment[]
  // Flux CDAR technique associé (§ 6.2) — fusionné dans l'affichage plutôt que
  // listé séparément (cf. maquette popin facture).
  afnor_flow: LifecycleEventAfnorFlow | null
}

export interface CreateLifecycleEventPayload {
  status: string
  reason?: string | null
  action?: string | null
  comment?: string | null
  confirmed?: boolean
}

export function getLifecycleCatalog(): Promise<LifecycleCatalog> {
  return apiFetch('/api/ihm/lifecycle-catalog', {}, 'Failed to load lifecycle catalog')
}

export function listLifecycleEvents(invoiceId: number): Promise<LifecycleEvent[]> {
  return apiFetch(`/api/ihm/invoices/${invoiceId}/lifecycle-events`, {}, 'Failed to list lifecycle events')
}

export function lifecycleEventAttachmentDownloadUrl(
  invoiceId: number,
  eventId: number,
  attachmentId: number,
): string {
  return `${API_BASE}/api/ihm/invoices/${invoiceId}/lifecycle-events/${eventId}/attachments/${attachmentId}/download`
}

export function createLifecycleEvent(
  invoiceId: number,
  payload: CreateLifecycleEventPayload,
): Promise<LifecycleEvent> {
  return apiFetch(
    `/api/ihm/invoices/${invoiceId}/lifecycle-events`,
    { method: 'POST', json: payload },
    'Failed to create lifecycle event',
  )
}

export interface RetryAfnorFlowPayload {
  reason?: string | null
  action?: string | null
  comment?: string | null
}

/** Renvoie un CDAR sortant resté en erreur — `overrides` absent : renvoi tel quel,
 * fourni : remplace motif/action/commentaire avant renvoi (§ fiche facture). */
export function retryAfnorFlow(
  invoiceId: number,
  flowId: number,
  overrides?: RetryAfnorFlowPayload,
): Promise<LifecycleEvent> {
  return apiFetch(
    `/api/ihm/invoices/${invoiceId}/afnor-flows/${flowId}/retry`,
    { method: 'POST', ...(overrides !== undefined && { json: overrides }) },
    'Failed to retry AFNOR flow',
  )
}

File: backend/app/schemas/lifecycle.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class LifecycleEventDetailRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    reason: str | None
    action: str | None
    comment: str | None


class LifecycleEventPaymentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    amount: float
    currency: str
    payment_date: date


class LifecycleEventAttachmentRead(BaseModel):
    """`has_file` : construit explicitement côté endpoint (jamais par `from_attributes`
    automatique) — dérivé de `file_path`, jamais le contenu du fichier lui-même."""

    id: int
    filename: str
    has_file: bool


class AfnorFlowRead(BaseModel):
    """`has_file` : idem, dérivé de `file_path is not None` sans jamais exposer le
    contenu (téléchargé via un endpoint dédié, cf. `invoices.py`)."""

    id: int
    flow_id: str | None
    direction: str
    flow_type: str
    syntax: str
    processing_rule: str | None
    state: str
    has_file: bool


class LifecycleEventRead(BaseModel):
    """Toujours construit explicitement (cf. `invoices.py`), jamais par
    `from_attributes` automatique : `attachments`/`payments` mêlent des sous-schémas
    qui, eux, dérivent un champ (`has_file`) absent du modèle ORM."""

    id: int
    invoice_id: int
    event_datetime: datetime
    status: str
    direction: str
    amount: float | None
    currency: str | None
    details: list[LifecycleEventDetailRead]
    payments: list[LifecycleEventPaymentRead]
    attachments: list[LifecycleEventAttachmentRead]
    # Flux CDAR technique associé (§ 6.2) — permet à l'IHM de fusionner en une seule
    # ligne le statut métier et son état de transmission, sans requête séparée sur
    # "Flux AFNOR" (cf. maquette popin facture).
    afnor_flow: AfnorFlowRead | None


class RetryAfnorFlow(BaseModel):
    """Corps optionnel du renvoi d'un `AfnorFlow` sortant en erreur (§ IHM fiche
    facture) : absent (`null`/corps vide), le CDAR est renvoyé tel quel ; fourni, ces
    valeurs remplacent motif/action/commentaire de l'événement avant renvoi."""

    reason: str | None = None
    action: str | None = None
    comment: str | None = None


class StatusCatalogEntry(BaseModel):
    key: str
    label: str
    cdar_code: str
    mdt88_code: str | None
    manual_side: str | None
    requires_detail: bool
    requires_confirmation: bool
    # Sous-ensemble de `reasons` accepté pour CE statut par SuperPDP (§ lifecycle_catalog
    # StatusInfo.allowed_reasons) — vide tant que `requires_detail` est faux.
    allowed_reasons: list[str] = Field(default_factory=list)


class LifecycleCatalogRead(BaseModel):
    statuses: list[StatusCatalogEntry]
    reasons: dict[str, str]
    actions: dict[str, str] = Field(default_factory=dict)
